src/walletd/runtime.rs

```diff
@@ -135,1192 +64,33 @@ impl esb::Handler<ServiceBus> for Runtime {
-                        .find(|vk| vk.tag() == &SharedKeyId::new(SHARED_VIEW_KEY_ID))
-                        .unwrap()
-                        .elem();
-                    let view = view_key_alice + view_key_bob;
-                    info!(
-                        "{} | Full secret monero view key: {}",
-                        swap_id.swap_id(),
-                        view.bright_green_bold()
-                    );
-                    let network = pub_offer.offer.network.into();
-                    let keypair = monero::KeyPair { view, spend };
-                    let corresponding_address = monero::Address::from_keypair(network, &keypair);
-                    info!(
-                        "{} | Corresponding address: {}",
-                        swap_id.swap_id(),
-                        corresponding_address.addr()
-                    );
-                    let address = self
-                        .xmr_addrs
-                        .remove(&get_swap_id(&source)?)
-                        .expect("checked at the start of a swap");
-
-                    endpoints.send_to(
-                        ServiceBus::Ctl,
-                        ServiceId::Wallet,
-                        ServiceId::Database,
-                        BusMsg::Ctl(CtlMsg::SetAddressSecretKey(AddressSecretKey::Monero {
-                            address: corresponding_address,
-                            secret_key_info: MoneroSecretKeyInfo {
-                                swap_id: Some(swap_id),
-                                spend: keypair.spend.as_bytes().try_into().unwrap(),
-                                view: keypair.view.as_bytes().try_into().unwrap(),
-                                creation_height: None,
-                            },
-                        })),
-                    )?;
-
-                    let sweep_keys = SweepMoneroAddress {
-                        source_view_key: view,
-                        source_spend_key: spend,
-                        destination_address: address,
-                        minimum_balance: pub_offer.offer.accordant_amount,
-                    };
-                    endpoints.send_to(
-                        ServiceBus::Ctl,
-                        ServiceId::Wallet,
-                        source,
-                        BusMsg::Ctl(CtlMsg::SweepAddress(SweepAddressAddendum::Monero(
-                            sweep_keys,
-                        ))),
-                    )?;
-                } else {
-                    error!("Call to refund transaction expects an Alice wallet");
+            CtlMsg::CreateSwapKeys(public_offer, wallet_token) => {
+                if wallet_token != self.wallet_token {
+                    return Err(Error::InvalidToken);
                 }
+                let wallet_index = self.node_secrets.increment_wallet_counter();
+                let key_manager = KeyManager::new(self.node_secrets.wallet_seed, wallet_index)?;
```

Comment:
Do we have access to the swap id here if we need to? I'd like to refactor the `KeyManager` to adjust this new usage (with the cache system) and more importantly to take the swap id instead of a counter (which is super insecure). We also can make use of the offer uuid but I'd prefer the swap id.

Reply:
I didn't check since a long time ago, but this counter is not persisted right, it resets at every boot? The seed is persisted which is correct, so if we use the same counter twice we generate the same keys.

Reply:
No, the counter is persisted as well.

Reply:
Ok, the counter is persisted on disk at each increment. Then you can ignore my comment.

But if we want to simplify this process and do not need to persist changes on disk every time we can remove the counter and use the swap id at manager instantiation.

Reply:
I'd rather have a deterministic seed than not. If there is ever a case where a user looses the funding address data, he could still re-create it from the wallet secret and the counter.

Reply:
What I propose is: seed persited on disk + swap id to generate the keys (instead of a counter). The user has no knowledge about the counter's value used, but he knows the swap id.

Reply:
This would mean that the user would have to keep track of all swap ids in perpetuity. For the counter, the user just needs the current count (a single integer) to generate all the keys, which is currently persisted in the same file as the secret.

So the current situation is:

key.dat: NodeSecrets { peerd_secret_key, wallet_seed, wallet_counter }

And here is what I think it will look like if we use the swap id instead:

key.dat: Node Secrets { peerd_secret_key, wallet_seed, Vec<SwapId> }

I'd prefer the first option, seems much safer to me.

Reply:
The wallet counter is incremented/updated for each new swap, and the last generated `u32` is persisted in the file.

https://github.com/farcaster-project/farcaster-node/blob/c92187fc65a561174b401619fc43b74f7e738d2d/src/walletd/runtime.rs#L299-L301

https://github.com/farcaster-project/farcaster-node/blob/c92187fc65a561174b401619fc43b74f7e738d2d/src/walletd/opts.rs#L148

If you do 10 swaps and you need the keys for the `#3`, if you know the correct order you can load the counter and go ask for the correct index. For me this seems harder than having access the the swap id.

I wouldn't store the swap ids in the file, just the seed. Swap id can be in databased if wanted (the history of all swaps) or the responsibility of the user.

Btw I'm fine with the current design, but I'd prefer using static seed + swap id instead of static seed + counter. I don't see strong security difference between the two, only UX.

Reply:
If you have still have access to the swap database, you will still have access to its corresponding keys anyway. So there is no point in using the swap id with the seed to recover anything, since you already have the data. If you use just the counter you still have redundancy in case something happens to the database.

On a different note: Didn't we decide to phase out the swap id anyway? I'm really having a hard time seeing the benefits here.